[RFC] Execute sql commands in the original order of the schema diff

m-vo opened the pull request:
The result from doctrine's schema diff is already ordered to prevent collisions - e.g. foreign keys get dropped before deleting the respective tables. Currently we ignore the order when grouping the commands and letting the user decide which ones to execute (add table, modify table, drop table, and so on). 

This PR adds tracking of the original order for execution.

m-vo committed:
execute sql commands in the original order of the schema diff

m-vo commented on installation-bundle/src/Database/Installer.php:
```diff
@@ -54,22 +59,22 @@ public function getCommands(): array
+            $this->connection->query($commands[$hash]);
         }
-
-        throw new \InvalidArgumentException(sprintf('Invalid hash: %s', $hash));
```
@leofeyer Do we need this exception? It's not catched and therefore will only ever throw on the first invalid hash. 

I left it out in the current version of the PR but can reintroduce it to either throw with a list of all invalid hashes or similarly to like it is now.

leofeyer replied:
As discussed in Mumble on November 22nd, we should check if there are invalid hashes before we even execute the first query and throw and exception.

leofeyer commented on installation-bundle/src/Database/Installer.php:
```diff
@@ -54,22 +59,22 @@ public function getCommands(): array
-            if (isset($commands[$hash])) {
-                $this->connection->query($commands[$hash]);
+        $commands = [];
+        array_map(function ($c) use (&$commands): void {
```
What exactly does this do?

m-vo replied:
This just flattens the ``$commands`` array from something like this

```PHP
[
    'CREATE' => [
        'hash1' => 'sql1',
        'hash2' => 'sql2'
    ],
    'ALTER_TABLE' => [
        'hash3' => 'sql3'
    ],
    //...
];
```
to something like this:

```PHP
[
    'hash1' => 'sql1',
    'hash2' => 'sql2'
    'hash3' => 'sql3'
    //...
];
```

ausi replied:
These four lines can probably be simplified to:

```php
$commands = array_reduce($this->commands, 'array_merge', []);
```

WDYT?

leofeyer commented:
@contao/developers We should discuss how far we want to take this.

1. Do we still want to allow every single query to be disabled? Or would a dry-run view suffice? I am thinking about having a command to update the schema in the future.

2. Can we simply use the commands generated by Doctrine instead of splitting everything into groups (`ALTER_DROP`, `ALTER_ADD` etc.) and re-building it again afterwards? Where would we add custom queries added via the `sqlCompileCommands` hook then?

asaage commented:
> Do we still want to allow every single query to be disabled?

I would appreciate that - especially for drop-queries.
Sometimes i create tables and/or columns for backup/migration purposes which shouldn't be deleted by the installtool.

m-vo commented:
>Can we simply use the commands generated by Doctrine instead of splitting everything into groups 

What shall we do with the hook then? Unfortunately these groups get passed to the hook as an argument and removing / changing the structure is kind of a BC break?

https://github.com/contao/contao/blob/c35ab7c8b61fb0e0a09a3ba11ed1bf3ff9220e08/installation-bundle/src/Database/Installer.php#L166

By the way: I don't like the hook at all and I don't think it should be possible to remove elements. There are (or should be better) mechanisms to prevent tables beeing dropped / columns beeing changed and so on. 

Imho (ideally - besides the limitations of Doctrine) everything should be part of the schema with no manual tweaking after the diff.

https://github.com/contao/contao/blob/c35ab7c8b61fb0e0a09a3ba11ed1bf3ff9220e08/installation-bundle/src/Database/Installer.php#L96-L102

leofeyer commented:
@m-vo Do you have time to complete this PR this week? 🙈

m-vo committed:
reintroduce exception when submitting invalid hashes
(now throws before executing anything)

m-vo commented:
Time, that overrated thing. :wink: 
Here you go. 

Just thinking: In case the exception gets thrown, the user will see the generic error screen (when not running in the dev environment). Maybe we should instead throw a custom exception (e.g. `InvalidSqlHashEception` or `DatabaseMigrationException`), catch it and display a more user friendly message instead. Or use a more verbose error template in the install tool that includes the exception message. But that's a general thing and has not that much to do with this PR.

ausi commented on installation-bundle/src/Database/Installer.php:
```diff
@@ -165,9 +179,19 @@ private function compileCommands(): void
+            // make sure commands added via the hook are also appended to the command order
+            foreach ($return as $commandSet) {
+                foreach ($commandSet as $hash => $sql) {
+                    if (!isset($order[$hash])) {
```
Shouldn’t this be

```suggestion
                    if (!\in_array($hash, $order, true)) {
```

instead?

m-vo replied:
true! good catch

m-vo committed:
simplify flattening the command array

m-vo committed:
correctly add hashes from hooks

m-vo committed:
CS

ausi reviewed (approved):
Thank you for the quick fixes! ❤️

m-vo committed:
make sure to track all commands (even if they get split)

m-vo commented:
I just realized, that there was still something wrong. I didn't realized we were splitting the commands if there are multiple rows changing per table (see for loop in the huge switch statement). The last changes fix this. There is probably a way to simplify this but I'll need some sleep now - cannot concentrate anymore :smiley:.

leofeyer committed:
Fix the coding style

leofeyer commented:
Thank you @m-vo.

leofeyer commented:
Unfortunately, the changes do not work. 😢

I have tried to upgrade from a Contao 3.5 database, which does work before 5cd2b409d274878ecc1b0bb872ad5e7a9c930629, however now it shows:

<img width="941" alt="" src="https://user-images.githubusercontent.com/1192057/49075411-0d982d00-f237-11e8-8692-bbf3e6fc10c5.png">

bytehead commented:
I had this error in every 3.5 I've updated so far (IMO not related to this change). Removing the index manually before the update should help.

leofeyer commented:
You do not need to remove anything. @Toflar has fixed the issue you are referring to. And this PR has re-introduced it.

@m-vo I have noticed that `$this->commandOrder` has 93 entries but `$hashes` has 173. And since we run `array_intersect()`, we are skipping 80 queries! This cannot be right, can it?

m-vo commented:
Sounds like a bug. :/ I'll have a look at it this evening.

m-vo commented:
Found the problem and fixed it*. I've pushed it to my fork - shall I create a new PR? I've seen there have been some other changes in `checkEngineAndCollation` that are related I guess.

*) I totally overlooked that `checkEngineAndCollation` adds hashes as well (the array gets passed by reference), so I added those as well. There were some cases where this method added commands to the `ALTER TABLE` group that had already been in there in another group. In this case there would be more entries in the `$order` - to prevent one query beeing executed multiple times I therefore added an `array_unique` to strip all but the first occurence. <-- I wondered if having the sql command in there multiple times is intended (should it even run multiple times for some reason?) or a bug? 

Edit: Ah I see :-) that's exactly what has changed with 
```
                    if (isset($sql['ALTER_CHANGE'][$strKey])) {
                        unset($sql['ALTER_CHANGE'][$strKey]);
                    }
```

leofeyer commented:
> I've pushed it to my fork - shall I create a new PR?

Yes please.
